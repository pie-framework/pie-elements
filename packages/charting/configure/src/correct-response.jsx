import * as React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { Chart } from '@pie-lib/charting';
import isEqual from 'lodash/isEqual';
import isEmpty from 'lodash/isEmpty';

import Typography from '@material-ui/core/Typography';

const styles = (theme) => ({
  container: {
    marginBottom: theme.spacing.unit * 2.5,
    display: 'flex',
    flex: 1,
  },
  button: {
    marginTop: theme.spacing.unit * 3,
    cursor: 'pointer',
    background: '#eee',
    padding: theme.spacing.unit * 2,
    width: 'fit-content',
    borderRadius: '4px',
  },
  column: {
    flex: 1,
  },
  chartError: {
    border: `2px solid ${theme.palette.error.main}`,
  },
  errorText: {
    fontSize: theme.typography.fontSize - 2,
    color: theme.palette.error.main,
    paddingTop: theme.spacing.unit,
  },
  title: {
    marginBottom: theme.spacing.unit,
  },
});

const addCategoryProps = (correctAnswer, data) =>
  correctAnswer.map((correct, index) => ({
    ...correct,
    editable: index < data.length ? data[index].editable : true,
    interactive: index < data.length ? data[index].interactive : true,
    deletable: index >= data.length ? true : false,
  }));

const updateCorrectResponseData = (correctAnswer, data) => {
  if (!correctAnswer) {
    return data;
  }

  const correctAnswerData = [...correctAnswer];
  let correctResponseDefinition = [];

  (data || []).forEach((category, currentIndex) => {
    const editable = category.editable;
    const interactive = category.interactive;
    const label = editable && correctAnswer[currentIndex]?.label ? correctAnswer[currentIndex].label : category.label;
    const value =
      (interactive && correctAnswer[currentIndex]?.value) || (interactive && correctAnswer[currentIndex]?.value == 0)
        ? correctAnswer[currentIndex].value
        : category.value;

    correctResponseDefinition[currentIndex] = {
      label: label,
      value: value,
      editable: category.editable,
      interactive: category.interactive,
    };
  });

  if (correctResponseDefinition.length < correctAnswer.length) {
    const missingCategories = (correctAnswerData || []).slice(correctResponseDefinition.length, correctAnswer.length);

    return addCategoryProps((correctResponseDefinition || []).concat(missingCategories), data);
  }

  return correctResponseDefinition;
};

const insertCategory = (correctAnswer, data) => {
  const positionToInsert = data.length - 1;
  const { editable, interactive, deletable, ...categoryToInsert } = data[data.length - 1];

  (correctAnswer || []).splice(positionToInsert, 0, categoryToInsert);
  const correctAnswerData = [...correctAnswer];

  return addCategoryProps(correctAnswerData, data);
};

const removeCategory = (correctAnswer, data, positionToRemove) => {
  (correctAnswer || []).splice(positionToRemove, 1);
  const correctAnswerData = [...correctAnswer];

  return addCategoryProps(correctAnswerData, data);
};

export class CorrectResponse extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    model: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired,
    charts: PropTypes.array,
  };

  constructor(props) {
    super(props);
    this.state = {
      categories: updateCorrectResponseData(props.model.correctAnswer.data, props.model.data),
    };
  }

  changeData = (data) => {
    const { model, onChange } = this.props;
    const { correctAnswer } = model || {};

    onChange({
      ...model,
      correctAnswer: {
        ...correctAnswer,
        data: data.map(({ interactive, editable, index, ...keepAttrs }) => keepAttrs),
      },
    });
  };

  UNSAFE_componentWillReceiveProps(nextProps) {
    const { model: { data: nextData = [], correctAnswer: { data: nextCorrectAnswerData = [] } } = {} } = nextProps;
    const { model: { data = [] } = {} } = this.props;
    const { categories } = this.state;
    let nextCategories = [];

    if (nextData.length > data.length) {
      nextCategories = insertCategory(nextCorrectAnswerData, nextData);
    }

    if (nextData.length < data.length) {
      let removedIndex = nextData.length;

      // we need to remove the category from the correct answer data and categories, from the same index it was removed from the data
      // index is a property of the nextData category
      for (let index = 0; index < nextData.length; index++) {
        if (nextData[index].index !== index) {
          removedIndex = index;
          break;
        }
      }

      (this.props.model.correctAnswer.data || []).splice(removedIndex, 1);
      nextCategories = removeCategory(categories, nextData, removedIndex);
    }

    if (!isEqual(nextCorrectAnswerData, this.props.model.correctAnswer.data)) {
      nextCategories = nextCorrectAnswerData.map((correct, index) => ({
        ...correct,
        editable: index < data.length ? data[index].editable : true,
        interactive: index < data.length ? data[index].interactive : true,
      }));
    } else if (isEmpty(nextCategories)) {
      nextCategories = updateCorrectResponseData(nextCorrectAnswerData, nextData);

      nextCorrectAnswerData.map((answer, currentIndex) => {
        const dataExists = currentIndex < nextData.length;
        let label;
        let value;
        let deletable;

        if (dataExists) {
          label = nextData[currentIndex].editable ? answer.label : nextData[currentIndex].label;
          value = nextData[currentIndex].interactive ? answer.value : nextData[currentIndex].value;
        } else {
          label = answer.label;
          value = answer.value;
          deletable = true;
        }

        nextCorrectAnswerData[currentIndex] = {
          label: label,
          value: value,
        };
      });
    }

    if (
      !isEqual(nextCategories, data) ||
      !isEqual(nextCorrectAnswerData, categories) ||
      !isEqual(nextCategories, categories) ||
      (isEmpty(nextCategories) && isEmpty(data))
    ) {
      this.setState({ categories: nextCategories });
    }
  }

  render() {
    const { classes, model, charts, error, correctAnswerErrors } = this.props;
    const { categories } = this.state;
    const { domain = {}, range = {} } = model || {};
    const { identicalError, categoriesError } = correctAnswerErrors || {};

    return (
      <div>
        <div className={classes.title}>Define Correct Response</div>
        <div className={classes.container}>
          <div className={classes.column} key="graph">
            <Typography component="div" type="body1">
              <span>Use the tools below to define the correct answer.</span>
            </Typography>

            <div
              key={`correct-response-graph-${model.correctAnswer.name}`}
              className={identicalError || categoriesError ? classes.chartError : {}}
            >
              <Chart
                chartType={model.chartType}
                size={model.graph}
                domain={domain}
                range={range}
                charts={charts}
                data={categories}
                title={model.title}
                onDataChange={(data) => this.changeData(data)}
                addCategoryEnabled={model.addCategoryEnabled}
                categoryDefaultLabel={model.categoryDefaultLabel}
                error={error}
              />
            </div>

            {(identicalError || categoriesError) && (
              <Typography component="div" type="body1" className={classes.errorText}>
                <span>{identicalError || categoriesError}</span>
              </Typography>
            )}
          </div>
        </div>
      </div>
    );
  }
}

export default withStyles(styles)(CorrectResponse);
