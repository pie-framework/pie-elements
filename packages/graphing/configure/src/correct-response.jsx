import * as React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { GraphContainer as Graph } from '@pie-lib/graphing';

import { set } from 'lodash';

const styles = theme => ({
  column: {
    flex: 1
  },
  graphingTools: {
    marginTop: theme.spacing.unit * 3,
    color: '#ababab'
  },
  availableTools: {
    marginTop: theme.spacing.unit * 3,
    marginBottom: theme.spacing.unit * 3,
    display: 'flex'
  },
  availableTool: {
    cursor: 'pointer',
    margin: theme.spacing.unit,
    padding: theme.spacing.unit,
    textTransform: 'capitalize'
  },
  selectedTool: {
    background: '#d8d8d8',
    border: '2px solid #ababab'
  },
  container: {
    border: '2px solid #ababab',
    borderRadius: '4px',
    padding: `${theme.spacing.unit * 2}px ${theme.spacing.unit * 4}px`,
    background: '#fafafa'
  },
  button: {
    marginTop: theme.spacing.unit * 3,
    cursor: 'pointer',
    background: '#eee',
    padding: theme.spacing.unit * 2,
    width: 'fit-content',
    borderRadius: '4px'
  }
});

export class CorrectResponse extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    model: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired,
    tools: PropTypes.arrayOf(
      PropTypes.shape({ type: PropTypes.string.isRequired })
    )
  };

  changeMarks = (key, marks) => {
    const { model, onChange } = this.props;

    set(model, `answers.${key}.marks`, marks);
    onChange(model);
  };

  changeToolbarTools = toolbarTools => {
    const { model, onChange } = this.props;
    model.toolbarTools = toolbarTools;

    onChange(model);
  };

  toggleToolBarTool = tool => {
    const { toolbarTools } = this.props.model;

    const index = toolbarTools.findIndex(t => tool.type === t);

    if (index >= 0) {
      const update = [...toolbarTools];
      update.splice(index, 1);
      this.changeToolbarTools(update);
    } else {
      this.changeToolbarTools([...toolbarTools, tool.type]);
    }
  };

  render() {
    const { classes, model, onChange, tools } = this.props;
    const { toolbarTools } = model;

    const t = tools.map(tool => {
      tool.toolbar = toolbarTools.some(t => t === tool.type);
      return tool;
    });

    return (
      <div>
        Define Correct Response
        <div className={classes.graphingTools}>
          GRAPHING TOOLS
          <div className={classes.availableTools}>
            {t.map(tool => {
              const selected = toolbarTools.find(t => t === tool.type);

              return (
                <div
                  key={tool.type}
                  className={classnames(
                    classes.availableTool,
                    selected && classes.selectedTool
                  )}
                  onClick={() => this.toggleToolBarTool(tool)}
                >
                  {tool.type.toUpperCase()}
                </div>
              );
            })}
          </div>
        </div>
        <div className={classes.container}>
          {Object.keys(model.answers).map(mark => (
            <div key={`correct-response-graph-${model.answers[mark].name}`}>
              <p>{model.answers[mark].name}</p>

              <Graph
                size={{
                  width: model.graph && model.graph.width,
                  height: model.graph && model.graph.height
                }}
                domain={model.domain}
                range={model.range}
                title={model.title}
                labels={model.labels}
                marks={model.answers[mark].marks}
                backgroundMarks={model.backgroundMarks}
                onChangeMarks={marks => this.changeMarks(mark, marks)}
                tools={tools}
                currentTool={tools.length && tools[0].Component.type}
                defaultTool={tools.length && tools[0].type}
              />
            </div>
          ))}

          <div
            className={classes.button}
            onClick={() => {
              set(
                model,
                `answers.${`alternate${Object.keys(model.answers).length}`}`,
                {
                  name: `Alternate ${Object.keys(model.answers).length}`,
                  marks: []
                }
              );
              onChange(model);
            }}
          >
            ADD ALTERNATE
          </div>
        </div>
      </div>
    );
  }
}

export default withStyles(styles)(CorrectResponse);
