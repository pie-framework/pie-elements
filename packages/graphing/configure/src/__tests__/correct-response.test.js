import * as React from 'react';
import { shallow, mount } from 'enzyme';

import { Tools, CorrectResponse } from '../correct-response';
import defaultValues from '../defaults';

describe('Tools', () => {
  describe('renders', () => {
    it('snapshot', () => {
      const props = {
        classes: {},
        toolbarTools: [],
        toggleToolBarTool: jest.fn(),
      };

      expect(shallow(<Tools {...props} />)).toMatchSnapshot();
    });
  });
});

describe('CorrectResponse', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      classes: {},
      model: defaultValues.model,
      onChange: jest.fn(),
      tools: [],
    };

    wrapper = (newProps) => {
      const configureProps = { ...props, newProps };

      return mount(<CorrectResponse {...configureProps} />);
    };
  });

  describe('renders', () => {
    it('snapshot', () => {
      expect(wrapper()).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    it('changeMarks calls onChange', () => {
      const component = wrapper();
      const marks = [{ x: 1, y: 1, type: 'point' }];

      component.instance().changeMarks('alternateTest', marks);

      expect(component.instance().props.onChange).toBeCalledWith({
        ...defaultValues.model,
        answers: {
          ...defaultValues.model.answers,
          alternateTest: {
            marks,
          },
        },
      });
    });

    it('changeToolbarTools calls onChange', () => {
      const component = wrapper();
      component.instance().changeToolbarTools([]);

      expect(component.instance().props.model.toolbarTools).toEqual([]);
      expect(component.instance().props.onChange).toHaveBeenCalledWith({
        ...defaultValues.model,
        toolbarTools: [],
      });
    });
  });

  it('toggleToolBarTool calls onChange', () => {
    const component = wrapper();
    component.instance().toggleToolBarTool('point');

    const { props } = component.instance();

    expect(props.onChange).toHaveBeenCalledWith({ ...defaultValues.model, toolbarTools: ['point'] });
  });

  it('addAlternateResponse calls onChange', () => {
    const component = wrapper();
    component.instance().addAlternateResponse();

    const { props } = component.instance();

    const answers = props.model.answers;

    answers.alternate1 = { marks: [], name: 'Alternate 1' };

    expect(props.model.answers).toEqual(answers);
    expect(props.onChange).toHaveBeenCalledWith({ ...defaultValues.model, answers });
  });
});

describe('CorrectResponse: if answers is null it should still work as expected', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      classes: {},
      model: defaultValues.model,
      onChange: jest.fn(),
      tools: [],
    };

    props.model.answers = null;

    wrapper = (newProps) => {
      const configureProps = { ...props, newProps };

      return mount(<CorrectResponse {...configureProps} />);
    };
  });

  describe('renders', () => {
    it('snapshot', () => {
      expect(wrapper()).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    it('changeMarks calls onChange', () => {
      const component = wrapper();
      const marks = [{ x: 1, y: 1, type: 'point' }];

      component.instance().changeMarks('alternateTest', marks);

      expect(component.instance().props.onChange).toBeCalledWith({
        ...defaultValues.model,
        answers: {
          ...defaultValues.model.answers,
          alternateTest: {
            marks,
          },
        },
      });
    });

    it('changeToolbarTools calls onChange', () => {
      const component = wrapper();
      component.instance().changeToolbarTools([]);

      expect(component.instance().props.model.toolbarTools).toEqual([]);
      expect(component.instance().props.onChange).toHaveBeenCalledWith({
        ...defaultValues.model,
        toolbarTools: [],
      });
    });
  });

  it('toggleToolBarTool calls onChange', () => {
    const component = wrapper();
    component.instance().toggleToolBarTool('point');

    const { props } = component.instance();

    expect(props.onChange).toHaveBeenCalledWith({ ...defaultValues.model, toolbarTools: ['point'] });
  });

  it('addAlternateResponse calls onChange', () => {
    const component = wrapper();
    component.instance().addAlternateResponse();

    const { props } = component.instance();

    const answers = props.model.answers;

    answers.alternate1 = { marks: [], name: 'Alternate 1' };

    expect(props.model.answers).toEqual(answers);
    expect(props.onChange).toHaveBeenCalledWith({ ...defaultValues.model, answers });
  });
});

describe('CorrectResponse: if answers is undefined it should still work as expected', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      classes: {},
      model: defaultValues.model,
      onChange: jest.fn(),
      tools: [],
    };

    props.model.answers = undefined;

    wrapper = (newProps) => {
      const configureProps = { ...props, newProps };

      return mount(<CorrectResponse {...configureProps} />);
    };
  });

  describe('renders', () => {
    it('snapshot', () => {
      expect(wrapper()).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    it('changeMarks calls onChange', () => {
      const component = wrapper();
      const marks = [{ x: 1, y: 1, type: 'point' }];

      component.instance().changeMarks('alternateTest', marks);

      expect(component.instance().props.onChange).toBeCalledWith({
        ...defaultValues.model,
        answers: {
          ...defaultValues.model.answers,
          alternateTest: {
            marks,
          },
        },
      });
    });

    it('changeToolbarTools calls onChange', () => {
      const component = wrapper();
      component.instance().changeToolbarTools([]);

      expect(component.instance().props.model.toolbarTools).toEqual([]);
      expect(component.instance().props.onChange).toHaveBeenCalledWith({
        ...defaultValues.model,
        toolbarTools: [],
      });
    });
  });

  it('toggleToolBarTool calls onChange', () => {
    const component = wrapper();
    component.instance().toggleToolBarTool('point');

    const { props } = component.instance();

    expect(props.onChange).toHaveBeenCalledWith({ ...defaultValues.model, toolbarTools: ['point'] });
  });

  it('addAlternateResponse calls onChange', () => {
    const component = wrapper();
    component.instance().addAlternateResponse();

    const { props } = component.instance();

    const answers = props.model.answers;

    answers.alternate1 = { marks: [], name: 'Alternate 1' };

    expect(props.model.answers).toEqual(answers);
    expect(props.onChange).toHaveBeenCalledWith({ ...defaultValues.model, answers });
  });
});
