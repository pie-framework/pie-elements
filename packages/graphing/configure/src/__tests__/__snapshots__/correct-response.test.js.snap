// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CorrectResponse renders snapshot 1`] = `
<div>
  <WithStyles(Typography)
    component="div"
    variant="subheading"
  >
    Define Tool Set and Correct Response
  </WithStyles(Typography)>
  <WithStyles(Typography)
    component="div"
    variant="body1"
  >
    Use this interface to choose which graphing tools students will be able to use, and to define the correct answer
  </WithStyles(Typography)>
  <Tools
    classes={Object {}}
    hasErrors={false}
    onDefaultToolChange={[Function]}
    toggleToolBarTool={[Function]}
    toolbarTools={
      Array [
        "circle",
        "line",
        "label",
        "parabola",
        "point",
        "polygon",
        "ray",
        "segment",
        "sine",
        "vector",
      ]
    }
  />
  <div>
    <div>
      Correct Answer
    </div>
  </div>
  <Root
    axesSettings={
      Object {
        "includeArrows": Object {
          "down": true,
          "left": true,
          "right": true,
          "up": true,
        },
      }
    }
    backgroundMarks={Array []}
    coordinatesOnHover={false}
    disabledLabels={true}
    disabledTitle={true}
    domain={
      Object {
        "axisLabel": "x",
        "labelStep": 1,
        "max": 5,
        "min": -5,
        "step": 1,
      }
    }
    draggableTools={true}
    labels={Object {}}
    marks={Array []}
    onChangeMarks={[Function]}
    onChangeTools={[Function]}
    range={
      Object {
        "axisLabel": "y",
        "labelStep": 1,
        "max": 5,
        "min": -5,
        "step": 1,
      }
    }
    showLabels={true}
    size={
      Object {
        "height": 500,
        "width": 500,
      }
    }
    title=""
    toolbarTools={
      Array [
        "circle",
        "line",
        "label",
        "parabola",
        "point",
        "polygon",
        "ray",
        "segment",
        "sine",
        "vector",
      ]
    }
  />
  <div
    onClick={[Function]}
  >
    ADD ALTERNATE
  </div>
  <Component
    open={false}
  />
</div>
`;

exports[`CorrectResponse: if answers is null it should still work as expected renders snapshot 1`] = `
<div>
  <WithStyles(Typography)
    component="div"
    variant="subheading"
  >
    Define Tool Set and Correct Response
  </WithStyles(Typography)>
  <WithStyles(Typography)
    component="div"
    variant="body1"
  >
    Use this interface to choose which graphing tools students will be able to use, and to define the correct answer
  </WithStyles(Typography)>
  <Tools
    classes={Object {}}
    hasErrors={false}
    onDefaultToolChange={[Function]}
    toggleToolBarTool={[Function]}
    toolbarTools={Array []}
  />
  <div
    onClick={[Function]}
  >
    ADD ALTERNATE
  </div>
  <Component
    open={false}
  />
</div>
`;

exports[`CorrectResponse: if answers is undefined it should still work as expected renders snapshot 1`] = `
<div>
  <WithStyles(Typography)
    component="div"
    variant="subheading"
  >
    Define Tool Set and Correct Response
  </WithStyles(Typography)>
  <WithStyles(Typography)
    component="div"
    variant="body1"
  >
    Use this interface to choose which graphing tools students will be able to use, and to define the correct answer
  </WithStyles(Typography)>
  <Tools
    classes={Object {}}
    hasErrors={false}
    onDefaultToolChange={[Function]}
    toggleToolBarTool={[Function]}
    toolbarTools={Array []}
  />
  <div
    onClick={[Function]}
  >
    ADD ALTERNATE
  </div>
  <Component
    open={false}
  />
</div>
`;

exports[`Tools renders snapshot 1`] = `
<div>
  <div>
    <span>
      GRAPHING TOOLS
    </span>
  </div>
  <div />
</div>
`;
